syntax = "proto3";

package detection;

option go_package = "github.com/gsy/lightstore/ml/proto;detectionpb";

// DetectionService provides beverage detection capabilities
service DetectionService {
  // Detect performs object detection on an image
  rpc Detect(DetectRequest) returns (DetectResponse);

  // HealthCheck returns server health status
  rpc HealthCheck(Empty) returns (HealthResponse);

  // GetModelInfo returns current model metadata
  rpc GetModelInfo(Empty) returns (ModelInfo);

  // SyncClasses fetches class names from catalog (called by server internally)
  rpc SyncClasses(SyncClassesRequest) returns (SyncClassesResponse);
}

message Empty {}

message DetectRequest {
  bytes image = 1;                    // JPEG/PNG image bytes
  string device_id = 2;               // Source device identifier
  float confidence_threshold = 3;     // Minimum confidence (default: 0.5)
  float iou_threshold = 4;            // NMS IoU threshold (default: 0.45)
}

message DetectResponse {
  repeated Detection detections = 1;
  string model_version = 2;
  float inference_time_ms = 3;
  string request_id = 4;
}

message Detection {
  string class_name = 1;              // Human-readable name (e.g., "Coca-Cola 330ml")
  string sku_id = 2;                  // SKU ID from catalog
  int32 class_id = 3;                 // Model class index
  float confidence = 4;
  BoundingBox bbox = 5;
}

message BoundingBox {
  float x1 = 1;                       // Top-left x (normalized 0-1)
  float y1 = 2;                       // Top-left y (normalized 0-1)
  float x2 = 3;                       // Bottom-right x (normalized 0-1)
  float y2 = 4;                       // Bottom-right y (normalized 0-1)
}

message HealthResponse {
  bool healthy = 1;
  string status = 2;
  bool model_loaded = 3;
  int64 uptime_seconds = 4;
}

message ModelInfo {
  string version = 1;
  string architecture = 2;            // e.g., "yolov8s"
  repeated string class_names = 3;
  int32 input_width = 4;
  int32 input_height = 5;
  string trained_at = 6;
  float mAP50 = 7;                    // Validation mAP@0.5
  float mAP50_95 = 8;                 // Validation mAP@0.5:0.95
}

message SyncClassesRequest {
  repeated ClassMapping classes = 1;
}

message ClassMapping {
  int32 class_id = 1;
  string sku_id = 2;
  string class_name = 3;
}

message SyncClassesResponse {
  bool success = 1;
  int32 class_count = 2;
}
